pkg/bgp/speaker/speaker.go

```diff
@@ -247,43 +247,53 @@ func (s *MetalLBSpeaker) notifyNodeEvent(op Op, nodeMeta metaGetter, podCIDRs *[
 	return nil
 }
 
-// OnAddNode notifies the Speaker of a new node.
-func (s *MetalLBSpeaker) OnAddNode(node *slim_corev1.Node, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Add, node, nodePodCIDRs(node), false)
+func (s *MetalLBSpeaker) SubscribeToLocalNodeResource(lnr agentK8s.LocalNodeResource) {
+	go eventLoop[*slim_corev1.Node](s, lnr, nodePodCIDRs)
 }
 
-func (s *MetalLBSpeaker) OnUpdateNode(oldNode, newNode *slim_corev1.Node, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Update, newNode, nodePodCIDRs(newNode), false)
+func (s *MetalLBSpeaker) SubscribeToLocalCiliumNodeResource(lcnr agentK8s.LocalCiliumNodeResource) {
+	go eventLoop[*ciliumv2.CiliumNode](s, lcnr, ciliumNodePodCIDRs)
 }
 
-// OnDeleteNode notifies the Speaker of a node deletion.
-//
-// When the speaker discovers the node that it is running on
-// is shuttig down it will send a BGP message to its peer
-// instructing it to withdrawal all previously advertised
-// routes.
-func (s *MetalLBSpeaker) OnDeleteNode(node *slim_corev1.Node, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Delete, node, nodePodCIDRs(node), true)
+type metaGetterObject interface {
+	k8sRuntime.Object
+	metaGetter
 }
 
-// OnAddCiliumNode notifies the Speaker of a new CiliumNode.
-func (s *MetalLBSpeaker) OnAddCiliumNode(node *ciliumv2.CiliumNode, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Add, node, ciliumNodePodCIDRs(node), false)
-}
-
-// OnUpdateCiliumNode notifies the Speaker of an update to a CiliumNode.
-func (s *MetalLBSpeaker) OnUpdateCiliumNode(oldNode, newNode *ciliumv2.CiliumNode, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Update, newNode, ciliumNodePodCIDRs(newNode), false)
-}
-
-// OnDeleteCiliumNode notifies the Speaker of a CiliumNode deletion.
-//
-// When the speaker discovers the node that it is running on
-// is shuttig down it will send a BGP message to its peer
-// instructing it to withdrawal all previously advertised
-// routes.
-func (s *MetalLBSpeaker) OnDeleteCiliumNode(node *ciliumv2.CiliumNode, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Delete, node, ciliumNodePodCIDRs(node), true)
+func eventLoop[T metaGetterObject](s *MetalLBSpeaker, res resource.Resource[T], nodePodCIDRs func(T) *[]string) {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
shouldn't this use the context passed to `New`? Otherwise only way to stop here is to stop the resource.

Reply:
It's the daemon context which ends up being passed in - changes such that it is passed to the subscription too.